### PR TITLE
Madeleine add subfolder lizardconnector

### PR DESCRIPTION
**Pull request by GeOdin**

### CODE REVIEW

File: libs/lizard-connector/lizard_connector/connector.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
+# coding=utf-8
+"""
+Connector to Lizard api (e.g. https://demo.lizard.net/api/v2) for python.
+
+Includes:
+- Datatypes (Rasters, Timeseries, Events, Assets), these are still a work in
+  progress and experimental. In later versions these contain:
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:7]: In later versions **they will** contain

File: libs/lizard-connector/lizard_connector/connector.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+try:
+    from urllib.parse import urlencode
+    from urllib.parse import urljoin
+    import urllib.request as urllib_request
+    from urllib.request import urlopen
+except ImportError:
+    from urllib import urlencode
+    from urlparse import urljoin
+    import urllib2 as urllib_request
+    from urllib2 import urlopen
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:26]: ``try/except`` for python ``2.x/3.x`` compatibility? If so, please add a comment

File: libs/lizard-connector/lizard_connector/connector.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+class InvalidUrlError(Exception):
+    pass
+
+
+class Connector(object):
+
+    def __init__(self, max_results=1000, username=None, password=None,
+                 all_pages=True):
+        """
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:43]: why are username and password optional? Can you connect without them?

File: libs/lizard-connector/lizard_connector/connector.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+                            login is used.
+            password (str): lizard-api password to log in. Without one no login
+                            is used.
+            all_pages (bool): when set to True, on download all pages are
+                              obtained. When set to False only the first
+                              page is obtained on get.
+        """
+        super(Endpoint, self).__init__(**kwargs)
+        self.endpoint = endpoint
+        base = base.strip(r'/')
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:188]: You want to treat backslashes as literals?

File: libs/lizard-connector/lizard_connector/connector.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        if self.use_header:
+            return {
+                "username": self.username,
+                "password": self.password
+            }
+        return {}
+
+
+class Endpoint(Connector):
+    max_results = 1000
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:171]: This seems to be a constant. Usually we use upper case for them.

File: libs/lizard-connector/lizard_connector/connector.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+    def __init__(self, endpoint, base="https://demo.lizard.net", **kwargs):
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:173]: maybe ``base_url`` would be a better name?

File: libs/lizard-connector/lizard_connector/connector.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    @property
+    def queries(self):
+        """
+        This property needs to be implemented for the class to function.
+        """
+        raise NotImplementedError
+
+    def endpoints(self):
+        """Lists all available endpoints for this datatype."""
+        return list(self.queries.keys())
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/connector.py:253]: also python 2.x/3.x compliance? In python 2.x ``dict.keys()`` is already a list

File: libs/lizard-connector/lizard_connector/parsers.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# coding=utf-8
+
+
+def parse_element(json_, element):
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/parsers.py:4]: I wouldn't use the name ``json_`` is an function signature.

File: libs/lizard-connector/lizard_connector/parsers.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        json (dict): json from the Lizard api parsed into a dictionary.
+        element (str): the element you wish to get.
+
+    Returns:
+        A list of all elements in the root of the results attribute.
+    """
+    return [x[element] for x in json_]
+
+
+def parse_uuid(json_, endpoint=None):
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/parsers.py:18]: see comment above

File: libs/lizard-connector/lizard_connector/parsers.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    Get a list of a certain element from the root of the results attribute.
+
+    Args:
+        json (dict): json from the Lizard api parsed into a dictionary.
+        endpoint (str): endpoint you wish to query.
+
+    Returns:
+        A list of all uuid elements in the root of the results attribute.
+    """
+    uuid = 'unique_id' if endpoint == 'organisations' else 'uuid'
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/parsers.py:29]: ``uuid`` is also a python module, beter use somethin like ``unique_id``

File: libs/lizard-connector/lizard_connector/queries.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    """
+    return ','.join(str(x) for x in args)
+
+
+def wkt_point(lon, lat):
+    return ''.join(('POINT (', str(lon), ' ', str(lat), ')'))
+
+
+def wkt_polygon(polygon_coordinates):
+    points = [' '.join([str(x), str(y)]) for x, y in polygon_coordinates]
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:65]: why are using ``lon/lat`` for ``wkt_point()`` but ``x/y`` here? I would choose for one notation

File: libs/lizard-connector/lizard_connector/queries.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    Returns:
+        a query dictionary.
+    """
+    query = "geom_within" if endpoint == 'timeseries' else "in_bbox"
+    return QueryDictionary({query: bbox(south_west, north_east)})
+
+
+def distance_to_point(distance, lat, lon):
+    """
+    Query for records with distance meters from a given point.
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:122]: with distance **in** meters

File: libs/lizard-connector/lizard_connector/queries.py [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                                  (lat, lon).
+
+    Returns:
+        a query dictionary.
+    """
+    query = "geom_within" if endpoint == 'timeseries' else "in_bbox"
+    return QueryDictionary({query: bbox(south_west, north_east)})
+
+
+def distance_to_point(distance, lat, lon):
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:120]: I get why you use ``lat/lon`` (ISO standard for geographic point locations). But you should be aware that it is often assumed that this is just a different way of saying ``x/y``, which it is not.

File: libs/lizard-connector/lizard_connector/queries.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+def organisation(organisation_id=None, endpoint=None):
+    query_key = {
+        None: "",
+        "organisation": "",
+        "location": "organisation__"
+    }.get(endpoint, "location__organisation__") + "unique_id"
+
+    org_query = QueryDictionary()
+    if isinstance(organisation_id, str):
+        org_query.update({query_key: organisation_id})
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:167]: if ``organisation_id`` is an unicode string ``isinstance(x, str)`` will yield False. Beter use ``isinstance(x, basestring)``

File: libs/lizard-connector/lizard_connector/queries.py [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    elif organisation_id:
+        org_query.update({query_key: ','.join(org for org in organisation_id)})
+    return org_query
+
+
+def statistics(*statistics):
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:173]: missing docstring

File: libs/lizard-connector/lizard_connector/queries.py [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+    statistics = commaify(*[x if x != 'mean' else 'count,sum' for x in
+                            statistics])
+    return QueryDictionary(min_points=1, fields=statistics)
+
+
+def feature_info(lat, lng, layername):
+    return QueryDictionary(
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:180]: missing docstring

File: libs/lizard-connector/lizard_connector/queries.py [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+        agg='curve',
+        geom='POINT({lng}+{lat})'.format(lat=lat, lng=lng),
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:182]: earlier you used ``lon`` here you use ``lng``, keep it uniform

File: libs/lizard-connector/lizard_connector/queries.py [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+        srs='EPSG:4326',
+        raster_names=layername,
+        count=False
+    )
+
+
+def limits(layername, south_west, north_east):
+    return QueryDictionary(
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:190]: missing docstring

File: libs/lizard-connector/lizard_connector/queries.py [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        height=16,
+        srs='epsg:4326'
+    )
+
+
+def search(q):
+    return QueryDictionary(search=q)
+
+
+groundwater = QueryDictionary(object_type__id=107)
```
Comment (larsclaussen) [inline libs/lizard-connector/lizard_connector/queries.py:204]: what does this variable do?

**Comment by GeOdin** [comment]
@RoelvandenBerg:
Lars had some comments on the lizard-connector, perhaps you can look at them.

! Possible changes in the lizard-connector should not be made in this repo.
I'll look at the lizard-connector with Jackie when he gets back.

File: libs/lizard-connector/lizard_connector/connector.py (hunk 3 above)
Comment (RoelvandenBerg) [reply libs/lizard-connector/lizard_connector/connector.py:43]: yup: https://demo.lizard.net/api/v3/

**Comment by RoelvandenBerg** [comment]
Thanks @larsclaussen there are some valuable comments in there. I still have to take a look at the repo. When are you merging @GeOdin? I think I should make some improvements to the repo before you release it.

**Pull request closed without merging**